kani-compiler/src/codegen_cprover_gotoc/codegen/place.rs

```diff
@@ -642,8 +642,30 @@ impl<'tcx> GotocCtx<'tcx> {
         let projection =
             unwrap_or_return_codegen_unimplemented!(self, self.codegen_place_stable(place));
         if self.use_thin_pointer_stable(place_ty) {
-            // Just return the address of the place dereferenced.
-            projection.goto_expr.address_of()
+            // For ZST objects rustc does not necessarily generate individual objects.
+            let need_not_be_unique = match projection.goto_expr.value() {
+                ExprValue::Symbol { .. } => self.is_zst_stable(place_ty),
+                _ => false,
+            };
+            let address_of = projection.goto_expr.address_of();
+            if need_not_be_unique {
+                let global_zst_name = "__kani_zst_object";
+                let zst_typ = self.codegen_ty_stable(place_ty);
+                let global_zst_object = self.ensure_global_var(
+                    global_zst_name,
```

Comment:
will there be a name collision if there are multiple types of ZST that have their address taken?

Reply:
`ensure_global_var` will only ever create a single symbol of that name, so whichever ZST object is seen first will determine the type.

Reply:
I think this is misleading since the variable that is returned might have a different type with different alignment requirements. This could easily trigger some sort of sanity check later on.